File: src/main/kotlin/xyz/upperlevel/hgame/world/entity/CloseRangeAttack.kt
package xyz.upperlevel.hgame.world.entity

import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.physics.box2d.Contact
import com.badlogic.gdx.physics.box2d.Fixture
import com.badlogic.gdx.physics.box2d.FixtureDef
import com.badlogic.gdx.physics.box2d.PolygonShape
import xyz.upperlevel.hgame.world.FixtureSensor

class CloseRangeAttack(val player: Player) {
    private val leftSensor = Sensor()
    private val rightSensor = Sensor()

    val contact: Player?
        get() {
            return if (player.left) leftSensor.player else rightSensor.player
        }

    fun setupFixtures() {
        val leftSensorDef = FixtureDef().apply {
            shape = PolygonShape().apply {
                val w = Player.WIDTH / 4f
                val h = Player.HEIGHT / 2f
                setAsBox(w, h, Vector2(w, h), 0f)
            }
            isSensor = true
        }
        val rightSensorDef = FixtureDef().apply {
            shape = PolygonShape().apply {
                val w = Player.WIDTH / 4f
                val h = Player.HEIGHT / 2f
                setAsBox(w, h, Vector2(w * 3, h), 0f)
            }
            isSensor = true
        }
        player.body.createFixture(leftSensorDef).userData = leftSensor
        player.body.createFixture(rightSensorDef).userData = rightSensor
    }

    private class Sensor : FixtureSensor {
        var player: Player? = null
        var touchCount = 0

        override fun onTouchBegin(other: Fixture, contact: Contact) {
            val entity = other.body.userData
            if (entity == null || entity !is Player) return

            if (player == null) {
                player = entity
            } else if (player != entity) {
                throw RuntimeException("Too many players!")
            }
            touchCount++
        }

        override fun onTouchEnd(other: Fixture, contact: Contact) {
            val entity = other.body.userData
            if (entity == null || entity !is Player) return

            if (entity == player) {
                if (--touchCount == 0) {
                    player = null
                }
            }
        }
    }
}

File: game/src/entity/player/closeRangeAttack.ts
import {Player} from "./player";
// @ts-ignore
import * as planck from "planck-js"
import Fixture = planck.Fixture;
import FixtureDef = planck.FixtureDef;


class Sensor {
    // a lazy deletion list should be more performant... who cares
    colliders: Player[] = [];

    onTouchBegin(other: Fixture) {
        const entity = other.getUserData();

        if (entity == null) return;

        this.colliders.push(entity);
    }

    onTouchEnd(other: Fixture) {
        const entity = other.getUserData();

        if (entity == null) return;

        const index = this.colliders.indexOf(entity);

        if (index < 0) return;

        this.colliders.splice(index, 1);
    }
}

export class CloseRangeAttack {
    owner: Player;
    leftSensor = new Sensor();
    rightSensor = new Sensor();

    constructor(player: Player) {
        this.owner = player;

        const w = Player.WIDTH / 4;
        const h = Player.HEIGHT / 2;

        const leftSensorDef: FixtureDef = {
            shape: planck.Box(w, h, planck.Vec2(-w, h), 0),
            isSensor: true,
            userData: this.leftSensor,
            // Collide ONLY with players
            filterMaskBits: Player.COLLISION_CATEGORY,
        };

        const rightSensorDef: FixtureDef = {
            shape: planck.Box(w, h, planck.Vec2(w, h), 0),
            isSensor: true,
            userData: this.rightSensor,
            // Collide ONLY with players
            filterMaskBits: Player.COLLISION_CATEGORY,
        };

        player.body.createFixture(leftSensorDef);
        player.body.createFixture(rightSensorDef);
    }

    getContacts(): Player[] {
        return this.owner.flipX ? this.leftSensor.colliders : this.rightSensor.colliders;
    }
}
